pkg/upstream/cluster/loadbalancer.go

```diff
@@ -109,3 +110,107 @@ func (l *roundRobinLoadBalancer) ChooseHost(context types.LoadBalancerContext) t
 
 	return selectedHost
 }
+
+/*
+SW (Smooth Weighted) is a struct that contains weighted items and provides methods to select a weighted item.
+It is used for the smooth weighted round-robin balancing algorithm. This algorithm is implemented in Nginx:
+https://github.com/phusion/nginx/commit/27e94984486058d73157038f7950a0a36ecc6e35.
+Algorithm is as follows: on each peer selection we increase current_weight
+of each eligible peer by its weight, select peer with greatest current_weight
+and reduce its current_weight by total number of weight points distributed
+among peers.
+In case of { 5, 1, 1 } weights this gives the following sequence of
+current_weight's: (a, a, b, a, c, a, a)
```

Comment:
yes, following  details in each cycle is omitted, i will add them for easy understanding

```golang
/*
In case of { 5, 1, 1 } weights this gives the following sequence of
current_weight's:
     a  b  c
     0  0  0  (initial state)

     5  1  1  (a selected)
    -2  1  1

     3  2  2  (a selected)
    -4  2  2

     1  3  3  (b selected)
     1 -4  3

     6 -3  4  (a selected)
    -1 -3  4

     4 -2  5  (c selected)
     4 -2 -2

     9 -1 -1  (a selected)
     2 -1 -1

     7  0  0  (a selected)
     0  0  0
*/
```